### ISSUE TITLE
Missing non-bpf alternative for create_program_address

### ISSUE THREAD

**Issue by hendrikhofstadt** [solana-labs/solana#11458]
#### Problem

In 1.3.0 `create_program_address` was migrated to a syscall. However there is no non-syscall fallback for tests.

**Comment by jackcmay** [comment]
Good point, you are talking about here?
 https://github.com/solana-labs/solana/blob/master/sdk/src/program_stubs.rs

**Comment by hendrikhofstadt** [comment]
I actually also need to access it from the CLI code so I would appreciate if it was available outside of the `program` feature flag.

**Comment by jackcmay** [comment]
Yeah, I'm looking at that exact thing, easier said than done

**Comment by jackcmay** [comment]
Why does you CLI define `program`?

**Comment by hendrikhofstadt** [comment]
The CLI does not but I still need access to the derivation

**Comment by jackcmay** [comment]
Sorry, confused, if CLi doesn't define the `program` feature why can't you call it directly?

**Comment by hendrikhofstadt** [comment]
The CLI uses predefined derivation methods from the program which uses the syscall function.

**Comment by mvines** [comment]
@jackcmay, I found it weird that there's `Pubkey::create_program_address()` for non-programs and then `program::create_program_address` for (currently only) bpf programs.
https://github.com/solana-labs/solana/blob/7e25130529991b9420bf671bb396a27a888a4d54/sdk/src/pubkey.rs#L94-L116

**Comment by jackcmay** [comment]
@hendrikhofstadt Do those pre-defined derivation calls need to be behind the `program` feature in your program crate?

@mvines `program::crate_program_address` is only intended to be called from within the BPF sandbox by a program and whose purpose is to create an unresolved symbol in the shared object that gets resolved at load time.  `Pubkey::create_program_address() is should not be called directly from a program due to is computational complexity and dependencies.  Did you have something else in mind?

**Comment by mvines** [comment]
> @mvines program::crate_program_address is only intended to be called from within the BPF sandbox by a program and whose purpose is to create an unresolved symbol in the shared object that gets resolved at load time. `Pubkey::create_program_address() is should not be called directly from a program due to it's computational complexity and dependencies. Did you have something else in mind?

I was just confused to find basically same function in two locations.

**Comment by hendrikhofstadt** [comment]
@jackcmay They don't because I want to share them between the CLI and the program.

Ideally they would resolve to the syscall in bpf and the Pubkey:: function in the CLI.
I could implement that myself with flags but would prefer if that was done right in the SDK.

**Comment by jackcmay** [comment]
Same signature, different function :-). One is just a wrapper

**Comment by jackcmay** [comment]
@hendrikhofstadt What do you see as the right way in the SDK?

In your program you can do:

```rust
    #[cfg(not(feature = "program"))]
    use solana_sdk::pubkey::create_program_address
    #[cfg(feature = "program")]
    use solana_sdk::program::create_program_address
```

Are you suggesting putting the stub also in `pubkey.rs`?

**Comment by hendrikhofstadt** [comment]
From a developer experience standpoint I think it would be more intuitive to have the function in just one place using and a target config flag to pick the implementation as you proposed.

**Comment by jackcmay** [comment]
@mvines @hendrikhofstadt 

wdyt: https://github.com/solana-labs/solana/pull/11460

**Comment by hendrikhofstadt** [comment]
@jackcmay Actually I think that syscall should rather do `find_program_address` or there should be a separate syscall.

The DX of having to precalculate nonces and pass them into the program is a bit painful. E.g. I have an instruction that uses 3 different derived accounts. In that case I have to pass in 3 different nonces and pass them through until the point where they are used.

I do understand that this negatively affects performance. At the same time why must derived keys not lie on the curve? Shouldn't the security of the curve be enough to guarantee that there will not be any collisions?

**Comment by jackcmay** [comment]
Program addresses are never meant to be signable by an outside entity so we split the domain so that wasn't possible.

`find_program_address` is a nicety to search for a valid program address, this could be done directly by your program using the `create_program_address`.  How about we continue this topic on discord.

What do you think about closing this issue as part of #11460 given the problem description?

**Comment by hendrikhofstadt** [comment]
@jackcmay agree. Let's continue on discord

Fixed by #11460
